hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -426,15 +428,30 @@ void notifyCallTimeout() {
             complete(CALL_TIMEOUT);
             return;
         }
+
         if (context.logger.isFinestEnabled()) {
-            context.logger.finest("Call timed-out either in operation queue or during wait-notify phase, retrying call: "
-                    + this);
+            context.logger.finest("Call timed-out either in operation queue or during wait-notify phase, retrying call: " + this);
         }
 
-        // decrement wait-timeout by call-timeout
-        long waitTimeout = op.getWaitTimeout();
-        waitTimeout -= callTimeoutMillis;
-        op.setWaitTimeout(waitTimeout);
+        long oldWaitTimeout = op.getWaitTimeout();
+        long newWaitTimeout;
+        if (oldWaitTimeout < 0) {
+            // The old wait-timeout is unbound; so the new wait-timeout will remain unbound.
+            newWaitTimeout = oldWaitTimeout;
+        } else {
+            // The old wait-timeout was bound.
+            // We need to subtract the elapsed time so that the waitTimeout gets smaller on every retry.
+
+            // first we determine elapsed time. To prevent of the elapsed time being negative due to cluster-clock reset, and
+            // the call timeout increasing instead of decreasing, the elapsedTime will be at least 0.
+            // For elapsed time we rely on cluster-clock, since op.invocationTime is also based on it.
+            long elapsedTime = max(0, context.clusterClock.getClusterTime() - op.getInvocationTime());
+
+            // We need to take care of not running into a negative wait-timeout, because it will be interpreted as infinite.
+            // That why the max with 0, so that 0 is going to be the smallest remaining timeout
+            newWaitTimeout = max(0, oldWaitTimeout - elapsedTime);
```

Comment:
```
  /**
     * Returns the wait timeout in millis. -1 means infinite wait, and 0 means no waiting at all.
     *
     * The wait timeout is the amount of time a {@link BlockingOperation} is allowed to parked in the
     * {@link com.hazelcast.spi.impl.operationparker.OperationParker}.
     *
     * Examples:
     * <ol>
     *     <li>in case of ILock.tryLock(10, ms), the wait timeout is 10 ms</li>
     *     <li>in case of ILock.lock(), the wait timeout is -1</li>
     *     <li>in case of ILock.tryLock(), the wait timeout is 0.</li>
     * </ol>
     *
     * The waitTimeout only has meaning for blocking operations. For non blocking operations the value is undefined.
     *
     * @return the wait timeout.
     */
    public final long getWaitTimeout() {
        return waitTimeout;
    }
```